teamcity: failed test: TestBatchTimeout

The following tests appear to have failed on master (testrace): TestBatchTimeout/WithTimeout, TestBatchTimeout

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestBatchTimeout).

[#1333232](https://teamcity.cockroachdb.com/viewLog.html?buildId=1333232):


```
TestBatchTimeout
--- FAIL: test/TestBatchTimeout (0.020s)
None



TestBatchTimeout
--- FAIL: testrace/TestBatchTimeout (0.020s)




TestBatchTimeout/WithTimeout
--- FAIL: test/TestBatchTimeout: TestBatchTimeout/WithTimeout (0.010s)

------- Stdout: -------
        batcher_test.go:330: 
            	Error Trace:	batcher_test.go:330
            	Error:      	Should be true
            	Test:       	TestBatchTimeout/WithTimeout



TestBatchTimeout/WithTimeout
--- FAIL: testrace/TestBatchTimeout: TestBatchTimeout/WithTimeout (0.010s)

------- Stdout: -------
        batcher_test.go:330: 
            	Error Trace:	batcher_test.go:330
            	Error:      	Should be true
            	Test:       	TestBatchTimeout/WithTimeout




```

Please assign, take a look and update the issue accordingly.

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -880,15 +880,18 @@ func (ctx *Context) runHeartbeat(
 
 			var response *PingResponse
 			sendTime := ctx.LocalClock.PhysicalTime()
-			err := contextutil.RunWithTimeout(goCtx, "rpc heartbeat", ctx.heartbeatTimeout,
-				func(goCtx context.Context) error {
-					// NB: We want the request to fail-fast (the default), otherwise we won't
-					// be notified of transport failures.
-					var err error
-					response, err = heartbeatClient.Ping(goCtx, request)
-					return err
-				})
-
+			ping := func(goCtx context.Context) (err error) {
+				// NB: We want the request to fail-fast (the default), otherwise we won't
+				// be notified of transport failures.
+				response, err = heartbeatClient.Ping(goCtx, request)
+				return err
+			}
+			var err error
+			if ctx.heartbeatTimeout > 0 {
+				err = contextutil.RunWithTimeout(goCtx, "rpc heartbeat", ctx.heartbeatTimeout, ping)
+			} else {
+				err = ping(goCtx)
+			}
 			if err == nil {
 				err = errors.Wrap(
 					checkVersion(ctx.version, response.ServerVersion),

--- a/pkg/util/contextutil/context.go
+++ b/pkg/util/contextutil/context.go
@@ -119,9 +119,6 @@ var _ net.Error = TimeoutError{}
 func RunWithTimeout(
 	ctx context.Context, op string, timeout time.Duration, fn func(ctx context.Context) error,
 ) error {
-	if timeout <= 0 {
-		return fn(ctx)
-	}
 	ctx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 	err := fn(ctx)

